src/main/java/org/candlepin/model/ConsumerService.java

```diff
@@ -0,0 +1,178 @@
+
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+
+
+public class ConsumerService {
+    private static final Logger log = LoggerFactory.getLogger(ConsumerService.class);
+
+    @Inject private EntitlementCurator entitlementCurator;
+    @Inject private ConsumerCurator consumerCurator;
+    @Inject private DeletedConsumerCurator deletedConsumerCurator;
+    @Inject private PrincipalProvider principalProvider;
+    @Inject private PoolManager poolManager;
+    @Inject private IdentityCertificateCurator identityCertificateCurator;
+    @Inject private ContentAccessCertificateCurator contentAccessCertificateCurator;
+    @Inject private CertificateSerialCurator certificateSerialCurator;
+    @Inject private KeyPairCurator keyPairCurator;
+
+    @Transactional
+    public void delete(Consumer consumer) {
+        log.debug("Deleting consumer: {}", consumer);
+        this.bulkDelete(Collections.singletonList(consumer));
+    }
+
+    @Transactional
+    public void bulkDelete(List<Consumer> consumers) {
+        if (consumers == null || consumers.isEmpty()) {
+            return;
+        }
+
+        List<String> consumerIdsToDelete = new ArrayList<>(consumers.size());
+        List<String> consumerUuidsToDelete = new ArrayList<>(consumers.size());
+        List<String> idCertsToDelete = new ArrayList<>(consumers.size());
+        List<String> caCertsToDelete = new ArrayList<>(consumers.size());
+        List<Long> serialsToRevoke = new ArrayList<>(consumers.size());
+
+        for (Consumer consumer : consumers) {
+            log.info("Deleting consumer: {}", consumer);
+
+            consumerIdsToDelete.add(consumer.getId());
+            consumerUuidsToDelete.add(consumer.getUuid());
+
+            IdentityCertificate idCert = consumer.getIdCert();
+            if (idCert != null) {
+                idCertsToDelete.add(idCert.getId());
+                serialsToRevoke.add(idCert.getSerial().getId());
+            }
+
+            ContentAccessCertificate contentAccessCert = consumer.getContentAccessCert();
+            if (contentAccessCert != null) {
+                caCertsToDelete.add(contentAccessCert.getId());
+                serialsToRevoke.add(contentAccessCert.getSerial().getId());
+            }
+        }
+
+        List<Entitlement> entsToRevoke = this.entitlementCurator.listByConsumerUuids(consumerUuidsToDelete);
+        // We're about to delete these consumers; no need to regen/dirty their dependent
+        // entitlements or recalculate status.
+        this.poolManager.revokeEntitlements(entsToRevoke, false);
+
+        int deletedFacts = this.consumerCurator.bulkDeleteFactsOf(consumerIdsToDelete);
```

Comment:
While all of this is technically correct, this tells me that our schema is the problem here. Relying entirely on the ORM layer to drive data integrity is a meme and not being able to do bulk operations without having to reimplement part of the ORM because of that tether is not something I want to get into.

I would argue the fix here is not to have this class, but to make sure our foreign keys are setup such that our children tables are properly updated should their referent get deleted.